### DISCUSSION TITLE
Two same dependent resources in different namespaces

### DISCUSSION THREAD

**Discussion by AleksanderBrzozowski** [operator-framework/java-operator-sdk#1969] [General]
Hi, I wanted to ask what is the correct way to implement two dependent resources in different namespaces.

We want to create one dependent resource in primary namespace and a second dependent resource in `istio-system` namespace.

Furthermore, we use `BulkDependentResource` interface to be able to create 0 or 1 resources based on the config in primary resource.

here are code snippets:

```
@KubernetesDependent(namespaces = [Namespaces.ISTIO_SYSTEM])
class IstioSystemDependent :
    BulkDependentResource<EnvoyFilter, Api>,
    KubernetesDependentResource<EnvoyFilter, Api>(EnvoyFilter::class.java),
    Updater<EnvoyFilter, Api>,
    SecondaryToPrimaryMapper<EnvoyFilter> by Mappers.fromAnnotatons() {

    override fun desiredResources(primary: Api, context: Context<Api>): Map<String, EnvoyFilter> {
        if (condition()) {
            return emptyMap()
        }
        val envoyFilter = EnvoyFilterBuilder()
            .withMetadata(primary.toMetadata())
            .withSpec(primary.toSpec())
            .build()
        return mapOf(envoyFilter.metadata.name to envoyFilter)
    }

    override fun getSecondaryResources(primary: Api, context: Context<Api>): Map<String, EnvoyFilter> {
        val secondaryResources = context.getSecondaryResources(EnvoyFilter::class.java)
                         .filter { it.metadata.namespace == Namespaces.ISTIO_SYSTEM }
        return secondaryResources
            .associateBy { envoyFilter -> envoyFilter.metadata.name }
    }
    ...
```

And here is the second dependent resource, it is very similar, but it doesn't specify `namespaces` inside `KubernetesDependent` annotation, and the `getSecondaryResources` method is slightly different:

```
   override fun getSecondaryResources(primary: Api, context: Context<Api>): Map<String, EnvoyFilter> {
        val secondaryResources = context.getSecondaryResources(EnvoyFilter::class.java)
            .filter { it.metadata.namespace == primary.metadata.namespace }
        return secondaryResources
            .associateBy { envoyFilter -> envoyFilter.metadata.name }
    }
```

The question is if these filter methods are necessary here, or if there is another way to do this:

```
filter { it.metadata.namespace == Namespaces.ISTIO_SYSTEM }

filter { it.metadata.namespace == primary.metadata.namespace }
```

The problem is that these both resources use two event sources - EnvoyFilter for `istio-system` namespace and EnvoyFilter for primary resource namespace. As a result, this method: `context.getSecondaryResources(EnvoyFilter::class.java)` returns resources from both namespaces.

**Comment by csviri** [comment]
Hi @AleksanderBrzozowski 

> Furthermore, we use BulkDependentResource interface to be able to create 0 or 1 resources based on the config in primary resource.

This seems to me just a reconcilePrecondition for a standard dependent, no need for a bulk dependent.

> The question is if these filter methods are necessary here, or if there is another way to do this:

Did you check this one in the docs:

https://javaoperatorsdk.io/docs/dependent-resources#multiple-dependent-resources-of-same-type

that part thould cover this use case as far I can see.

**Reply by AleksanderBrzozowski** [reply]
@csviri 
Thanks for your response 🙂 

> This seems to me just a reconcilePrecondition for a standard dependent, no need for a bulk dependent.

I tried `reconcilePrecondition` and it works as expected. Our use case is that user provides the name of the resource that is going to be created, for example `bar`. However, it can be changed later from `bar` to `foo`, and in this case we need to delete old `bar` resource, and create new `foo` resource. I tried this scenario with `reconcilePrecondition` and I matched based on name, like:

```
override fun isMet(
    dependentResource: DependentResource<Dependent, Primary>,
    primary: Primary,
    context: Context<Primary>
): Boolean {
    val secondary = dependentResource.getSecondaryResource(primary, context)
        .orElse(null) ?: return true
    val expectedName = primary.toMetadata().name
    val actualName = secondary.metadata.name
    return expectedName == actualName
}
```

From what I checked, when there is `bar` dependent created, and we edit primary resource to change the name of dependent to `foo`, the workflow starts to reconcile and `isMet` returns `false`, thus dependent is deleted.

And after it is deleted, the workflow reconciles once again, and this time `isMet` returns true (because there is no dependent resource) and it creates new dependent with new name. Am I right? 🙂 

> Did you check this one in the docs:
> https://javaoperatorsdk.io/docs/dependent-resources#multiple-dependent-resources-of-same-type
> that part thould cover this use case as far I can see.

When I tried to use `ResourceDiscriminator`, I got an error:

```
java.lang.IllegalArgumentException: There are multiple EventSources registered for type io.fabric8.istio.api.networking.v1alpha3.EnvoyFilter
```

```
    override fun distinguish(
        resource: Class<EnvoyFilter>,
        primary: Api,
        context: Context<Primary>
    ): Optional<EnvoyFilter> {
        return context.getSecondaryResource(EnvoyFilter::class.java)
            .filter { it.metadata.namespace == primary.metadata.namespace }
    }
```

Do I need to use only one event source for both of these dependent resources? The method `context.getSecondaryResource(EnvoyFilter::class.java, <event source name>)` is deprecated.
